connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+                logger.warning(
+                    f"Retry count: {retry} out of {self.retry_count}. Exception: {exception}"
+                )
+                await asyncio.sleep(RETRY_INTERVAL**retry)
+
+    async def get_content(self, issue_key, attachment, timestamp=None, doit=False):
+        """Extracts the content for allowed file types.
+
+        Args:
+            issue_key (str): Issue key to generate `_id` for attachment document
+            attachment (dictionary): Formatted attachment document.
+            timestamp (timestamp, optional): Timestamp of attachment last modified. Defaults to None.
+            doit (boolean, optional): Boolean value for whether to get content or not. Defaults to False.
+
+        Returns:
+            dictionary: Content document with _id, _timestamp and attachment content
+        """
+        attachment_size = int(attachment["size"])
+        if not (self.enable_content_extraction and doit and attachment_size > 0):
+            return
+
+        attachment_name = attachment["filename"]
+        if os.path.splitext(attachment_name)[-1] not in TIKA_SUPPORTED_FILETYPES:
+            logger.debug(f"{attachment_name} can't be extracted")
+            return
+
+        if attachment_size > FILE_SIZE_LIMIT:
+            logger.warning(
+                f"File size {attachment_size} of file {attachment_name} is larger than {FILE_SIZE_LIMIT} bytes. Discarding the file content"
+            )
+            return
+
+        logger.debug(f"Downloading {attachment_name}")
+
+        document = {
+            "_id": f"{issue_key}-{attachment['id']}",
+            "_timestamp": attachment["created"],
+        }
+        temp_filename = ""
+        attachment_url = ATTACHMENT_CLOUD if self.is_cloud else ATTACHMENT_SERVER
+        async with NamedTemporaryFile(mode="wb", delete=False) as async_buffer:
+            async for response in self._api_call(
+                url_name=attachment_url,
+                attachment_id=attachment["id"],
+                attachment_name=attachment["filename"],
+            ):
+                async for data in response.content.iter_chunked(CHUNK_SIZE):
+                    await async_buffer.write(data)
+                temp_filename = async_buffer.name
+
+        logger.debug(f"Calling convert_to_b64 for file : {attachment_name}")
+        await asyncio.to_thread(convert_to_b64, source=temp_filename)
+        async with aiofiles.open(file=temp_filename, mode="r") as async_buffer:
+            # base64 on macOS will add a EOL, so we strip() here
+            document["_attachment"] = (await async_buffer.read()).strip()
+        await remove(temp_filename)
+        return document
+
+    async def ping(self):
+        """Verify the connection with Jira"""
+        self._validate_configuration()
+        self.session = self._generate_session()
+        if not self.ssl_disabled:
+            self.ssl_ctx = self._ssl_context(certificate=self.certificate)
```

Comment:
This should not happen in `ping` - it's not used in this function and `get_docs` depends on it. You should initialise it in such a way, that both `ping` and `get_docs` could work if called individually, not sequentially.

Reply:
Understood your concern here. But, the connector is designed in such a way that `ping` will always gets executed first and then the `get_docs` method. So, there would not be a chance of the `get_docs` method to be called individually since the framework itself is not allowing it. However, once we have a conclusion on this [comment](https://github.com/elastic/connectors-python/pull/493#issuecomment-1438452169), we'll update the code accordingly.

Reply:
It's not designed this way, it works this way now because it was the most convenient way to do. It might and will change in future, so we want to build connectors that are resilient to changes.

In short, you should treat `ping` and `get_docs()` as methods that can be called in any sequence OR even on their own, all these 3 cases are valid:

1. `ping(); get_docs()`
2. `ping(); ping()`
3. `get_docs()`

Reply:
Oh! got it. However this is related to this [comment](https://github.com/elastic/connectors-python/pull/493#issuecomment-1438452169) so do you want me to change this before getting the conclusion on it, or should wait for it? Till the time, i'll address other comments.

Reply:
I'll try to address the topic in the comment in the mean time, I'll try to get back to you asap

Reply:
Sure, please. Do let me know your opinion on this.

Reply:
Keeping this to both the places i.e. `ping` and `get_docs`